In [32]:
# Importing the libraries for data analysis
import pandas as pd

In [33]:
# Install xlrd to read excel files
#!pip install xlrd

In [34]:
# URL of the Excel file containing shark attack data
url = 'https://www.sharkattackfile.net/spreadsheets/GSAF5.xls'  


In [35]:
# Read the Excel file from the given URL into a DataFrame
df = pd.read_excel(url)

# Create a copy of the DataFrame to make changes
df_mod = df.copy()

# Now you can work with df_mod without affecting the original df

In [36]:
# Display the first few rows of the DataFrame
df_mod.head()

Date    Year        Type    Country              State  \
0  2024-09-16 00:00:00  2024.0  Unprovoked    Morocco   Southern Morocco   
1  2024-08-26 00:00:00  2024.0  Unprovoked    Jamaica        Montego Bay   
2  2024-08-06 00:00:00  2024.0  Unprovoked     Belize   Gulf of Honduras   
3  2024-07-23 00:00:00  2024.0  Unprovoked  Australia                NSW   
4  2024-07-18 00:00:00  2024.0  Unprovoked  Australia  Western Australia   

                            Location                     Activity  \
0                     West of Dakhla  Swimming - jumped off yacht   
1                           Falmouth                 Spearfishing   
2                    Central America           Swimming - Diving    
3  North Shore Beach, Port Macquarie                      Surfing   
4               Trigg beach Sterling                      Surfing   

                Name Sex Age  ...                        Species   \
0     German Tourist   F  30  ...        Reportedly a Great White   
1       Jahmari Reid   M  16  ...          Reportedly Tiger Shark   
2  Annabelle Carlson   F  15  ...           Reef shark 6ft (1.8m)   
3       Kai McKenzie   M  23  ...                  3m White Shark   
4       Ryan Lowther   M  41  ...  Undetermined small shall shark   

                                              Source  pdf href formula href  \
0         Andy Currie: Moroccan World News: Sky News  NaN          NaN  NaN   
1   Todd Smith: Daily Mail UK: Sky News: People .com  NaN          NaN  NaN   
2  Kevin McMurray Tracking Sharks.com New York Po...  NaN          NaN  NaN   
3          Nine.com.au: ABC.Net.Au Johannes Marchand  NaN          NaN  NaN   
4          Daily Mail: Sky News: The West Australian  NaN          NaN  NaN   

  Case Number Case Number.1 original order Unnamed: 21 Unnamed: 22  
0         NaN           NaN            NaN         NaN         NaN  
1         NaN           NaN            NaN         NaN         NaN  
2         NaN           NaN            NaN         NaN         NaN  
3         NaN           NaN            NaN         NaN         NaN  
4         NaN           NaN            NaN         NaN         NaN  

[5 rows x 23 columns]

In [37]:
# Show the shape of the DataFrame
print(f"The DataFrame has {df_mod.shape[0]} rows and {df_mod.shape[1]} columns.")

The DataFrame has 6970 rows and 23 columns.


In [38]:
# Display the information of the DataFrame
df_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  pdf             6799 non-null   object 
 16  href formula    6794 non-null   object 
 17  href            6796 non-null   o

In [39]:
# Check for missing values in the DataFrame
df_mod.isnull().sum()

Date                 0
Year                 2
Type                18
Country             50
State              482
Location           565
Activity           586
Name               220
Sex                579
Age               2995
Injury              35
Unnamed: 11        562
Time              3527
Species           3132
Source              19
pdf                171
href formula       176
href               174
Case Number        172
Case Number.1      173
original order     171
Unnamed: 21       6969
Unnamed: 22       6968
dtype: int64


## Analysis of DataFrame Columns

In this section, we will begin the analysis of the columns in the DataFrame. We will examine each column to understand its data, identify any necessary transformations, and clean the data as required. This process will help us prepare the data for further analysis and visualization.


### Analysis of the Date Column

In this section, we will analyze the `Date` column of the DataFrame. First, we will convert the `Date` column to a date data type to facilitate its manipulation and analysis. Then, we will explore the column to obtain information such as the date range and the frequency of events by year, month, etc.


In [9]:
# Show all the unique values in the 'Date' column
for date in df_mod['Date'].unique():
    print(date)

2024-09-16 00:00:00
2024-08-26 00:00:00
2024-08-06 00:00:00
2024-07-23 00:00:00
2024-07-18 00:00:00
2024-07-08 00:00:00
2024-07-05 00:00:00
2024-07-04 00:00:00
2024-06-07 00:00:00
2024-06-02 00:00:00
2024-05-28 00:00:00
2024-05-24 00:00:00
2024-05-23 00:00:00
2024-05-02 00:00:00
2024-04-26 00:00:00
2024-04-20 00:00:00
2024-04-08 00:00:00
15 Mar 2024
04 Mar 2024
02 Mar-2024
25 Feb-2024
14 Feb-2024
04-Feb-2024
29 Jan-2024
15 Jan-2024
09-Jan-2024
05-Jan-2024
30 Dec-2023
29 Dec-2023
28 Dec-2023
25 Dec-2023
24-Dec-2023
18 Dec-2023
14  Dec-2023
08 Dec-2023
04 Dec-2023
02 Dec-2023
30 Nov-2023
21 Nov-2023
10 Nov-2023
05 Nov-2023
03 Nov-2023
02 Nov-2023
Reported 02 Nov-2023
31 Oct-2023
25 Oct 2023
23 Oct-2023
21 Oct-2023
20 Oct 2023
15 Oct 2023
13 Oct-2023
02 Oct-2023
01 Oct-2023
18 Sep-2023
13 Sep-2023
12 Sep-2023
11 Sep-2023
09 Sep- 2023
Reported 06-Sep-2023
04 Sep 2023
25 Aug 2023
23 Aug-2023
22 Aug-2023
21 Aug-2023
16 Aug-2023
12 Aug-2023
07 Aug-2023
31 Jul-2023
30 Jul-2023
29 Jul-2023
24 J

In [10]:
import re
import datetime

# Step 3: Clean and Standardize Date Formats
def clean_date(date_str):
    # Check if the input is NaT (Not a Time)
    if pd.isna(date_str):
        return pd.NaT
    
    # Convert datetime objects to string
    if isinstance(date_str, (pd.Timestamp, datetime.datetime)):
        date_str = date_str.strftime('%d %b %Y')
    
    # Ensure the input is a string
    if not isinstance(date_str, str):
        return date_str
    
    # Remove words like 'Reported', 'Ca.', 'Early', 'Late', 'Between', 'Before', 'After', 'Anniversary Day', 'No date', and extra spaces
    date_str = re.sub(r'Reported|Ca\.|Ca|Early|Late|Between|Before|After|Anniversary Day|No date|During the war|Before the war|Said to be|World War II|a few years before|early|late|between|before|after|anniversary day|no date|during the war|before the war|said to be|world war ii', '', date_str, flags=re.IGNORECASE).strip()
    
    # Remove suffixes like '.a' and '.b'
    date_str = re.sub(r'\.\w$', '', date_str).strip()
    
    # Replace inconsistent formats
    date_str = re.sub(r'(\d{2})-(\w+)-(\d{4})', r'\1 \2 \3', date_str)  # 09-Sep-2023 -> 09 Sep 2023
    date_str = re.sub(r'(\d{2}) (\w+)-(\d{4})', r'\1 \2 \3', date_str)  # 09 Sep-2023 -> 09 Sep 2023
    date_str = re.sub(r'(\d{2})-(\w+)-(\d{2})', r'\1 \2 20\3', date_str)  # 09-Sep-23 -> 09 Sep 2023
    date_str = re.sub(r'(\w+)-(\d{2})-(\d{4})', r'\2 \1 \3', date_str)  # Aug-24-1806 -> 24 Aug 1806
    
    # Handle year ranges (e.g., 1900-1905)
    date_str = re.sub(r'(\d{4})-(\d{4})', lambda m: f'01 Jan {(int(m.group(1)) + int(m.group(2))) // 2}', date_str)  # 1900-1905 -> 1902
    
    # Handle dates with only year (e.g., 1900)
    if re.match(r'^\d{4}$', date_str):
        date_str = f'01 Jan {date_str}'  # Convert to 01 Jan 1900
    
    # Handle dates with month and year (e.g., Jan 1900)
    if re.match(r'^\w+ \d{4}$', date_str):
        date_str = f'01 {date_str}'  # Convert to 01 Jan 1900
    
    # Handle dates with month and year in different format (e.g., Sep-1805)
    if re.match(r'^\w+-\d{4}$', date_str):
        date_str = f'01 {date_str.replace("-", " ")}'  # Convert to 01 Sep 1805
    
    # Handle dates with month name and year (e.g., October 1815)
    if re.match(r'^\w+ \d{4}$', date_str):
        date_str = f'01 {date_str}'  # Convert to 01 October 1815
    
    # Handle dates with "or" (e.g., 1990 or 1991)
    if re.match(r'^\d{4} or \d{4}$', date_str):
        date_str = date_str.split(' or ')[0]  # Take the first year
    
    # Handle dates with "B.C." (e.g., Ca. 214 B.C.)
    date_str = re.sub(r'B\.C\.', 'BC', date_str, flags=re.IGNORECASE)
    
    # Handle dates with "A.D." (e.g., Ca. 77 A.D.)
    date_str = re.sub(r'A\.D\.', 'AD', date_str, flags=re.IGNORECASE)
    
    # Handle dates with "Circa" or "Ca." (e.g., Circa 1855)
    date_str = re.sub(r'Circa|circa', '', date_str).strip()
    
    # Specific cases mapping
    specific_cases = {
        r'Before (\d{4})': lambda m: f'01 Jan {int(m.group(1)) - 1}',
        r'Before (\d{2})-(\w+)-(\d{4})': lambda m: f'{int(m.group(1)) - 1} {m.group(2)} {m.group(3)}',
        r'Between (\d{4}) & (\d{4})': lambda m: f'01 Jan {(int(m.group(1)) + int(m.group(2))) // 2}'
    }
    
    # Apply specific cases
    for pattern, func in specific_cases.items():
        date_str = re.sub(pattern, func, date_str)
    
    return date_str

# Apply the cleaning function to the 'Date' column
df_mod['Date'] = df_mod['Date'].apply(clean_date)


In [11]:
# Step 4: Convert to Datetime and Verify Null Values
df_mod['Date'] = pd.to_datetime(df_mod['Date'], errors='coerce')
null_dates_count = df_mod['Date'].isnull().sum()
print(f"Number of null dates after conversion: {null_dates_count}")

Number of null dates after conversion: 415


In [12]:
# Step 5: Handle Null Values
# Option 1: Drop Rows with Null Values
df_mod = df_mod.dropna(subset=['Date'])
print(f"Number of null dates after dropping: {df_mod['Date'].isnull().sum()}")

Number of null dates after dropping: 0


In [13]:
# Step 6: Extract Year, Month, and Day
df_mod['Year'] = df_mod['Date'].dt.year
df_mod['Month'] = df_mod['Date'].dt.month
df_mod['Day'] = df_mod['Date'].dt.day

# Verify the transformations
print(df_mod[['Date', 'Year', 'Month', 'Day']].head(10))

        Date  Year  Month  Day
0 2024-09-16  2024      9   16
1 2024-08-26  2024      8   26
2 2024-08-06  2024      8    6
3 2024-07-23  2024      7   23
4 2024-07-18  2024      7   18
5 2024-07-08  2024      7    8
6 2024-07-05  2024      7    5
7 2024-07-04  2024      7    4
8 2024-07-04  2024      7    4
9 2024-07-04  2024      7    4


### Analysis of the Year Column

In this section, we will analyze the `Year` column of the dataset. We will examine the distribution of years, identify any anomalies, and clean the data as necessary. This analysis will help us understand the temporal trends in the dataset and prepare the data for further analysis and visualization.
```

In [14]:
# Extract unique values from the 'Year' column
unique_years = df_mod['Year'].unique()

# Show the unique years
print(f"Unique years: {unique_years}")

Unique years: [2024 2023 2022 2021 2020 2010 2019 2018 2017 2016 2015 2014 2013 2012
 2011 2008 2009 2006 2007 2005 2004 2003 2002 2001 2000 1999 1998 1997
 1996 1995 1984 1994 1993 1992 1991 1990 1989 1988 1987 1986 1985 1983
 1982 1962 1981 1980 1979 1978 1977 1976 1975 1974 1973 1972 1971 1970
 1969 1968 1967 1966 1965 1964 1954 1963 1961 1960 1959 1958 1957 1956
 1955 1953 1952 1951 1950 1949 1948 1848 1947 1946 1945 1944 1943 1942
 1941 1940 1939 1938 1937 1936 1935 1934 1933 1932 1931 1930 1929 1928
 1927 1926 1925 1924 1923 1922 1921 1920 1919 1918 1917 1916 1915 1914
 1913 1912 1911 1910 1909 1908 1907 1906 1905 1904 1903 1902 1901 1900
 1899 1898 1897 1896 1895 1894 1893 1892 1891 1890 1889 1888 1887 1886
 1885 1884 1883 1882 1881 1880 1879 1878 1877 1876 1875 1874 1873 1872
 1871 1870 1869 1868 1867 1866 1865 1864 1863 1862 1861 1860 1859 1858
 1857 1856 1855 1853 1852 1851 1850 1849 1847 1846 1845 1844 1842 1841
 1840 1839 1837 1836 1835 1834 1832 1831 1830 1829 1828 1827 18

In [15]:
# Show the first few rows of the 'Year' column
print(df_mod['Year'].head(10))

0    2024
1    2024
2    2024
3    2024
4    2024
5    2024
6    2024
7    2024
8    2024
9    2024
Name: Year, dtype: int32


In [16]:
# Count the frequency of each year
year_counts = df_mod['Year'].value_counts().sort_index()

# Display the frequency of each year
print(year_counts)

Year
1691      1
1703      1
1723      1
1733      1
1738      1
       ... 
2020     99
2021    107
2022     98
2023     97
2024     35
Name: count, Length: 237, dtype: int64


In [17]:
# Verify the data type of the 'Year' column
print(df_mod['Year'].dtype)

int32


In [18]:
# Convert the 'Year' column to a string
df_mod['Year'] = df_mod['Year'].astype(str)

# Verify the data type of the 'Year' column
print(df_mod['Year'].dtype)

object



### Analysis of the Type Column

In this section, we will analyze the `Type` column of the dataset. We will examine the different types of incidents recorded, identify any anomalies, and clean the data as necessary. This analysis will help us understand the nature of the incidents and prepare the data for further analysis and visualization.


In [19]:
# Extract unique values from the 'Type' column
unique_types = df_mod['Type'].unique()
print(f"Unique types: {unique_types}")

Unique types: ['Unprovoked' ' Provoked' 'Provoked' 'Questionable' 'Watercraft'
 'Sea Disaster' nan '?' 'Unconfirmed' 'Unverified' 'Invalid'
 'Under investigation' 'Boat']


In [20]:
# Step 1: Remove leading/trailing spaces
df_mod['Type'] = df_mod['Type'].str.strip()

In [21]:
# Step 2: Replace '?' and 'nan' with 'Unknown'
df_mod['Type'] = df_mod['Type'].replace(['?', 'nan'], 'Unknown')

In [22]:
# Step 3: Verify the cleaned unique values
unique_types_cleaned = df_mod['Type'].unique()
print(f"Cleaned unique types: {unique_types_cleaned}")

Cleaned unique types: ['Unprovoked' 'Provoked' 'Questionable' 'Watercraft' 'Sea Disaster' nan
 'Unknown' 'Unconfirmed' 'Unverified' 'Invalid' 'Under investigation'
 'Boat']


In [23]:
# Step 4: Count the frequency of each type
type_counts_cleaned = df_mod['Type'].value_counts()
print(type_counts_cleaned)

Type
Unprovoked             4821
Provoked                612
Invalid                 519
Watercraft              337
Sea Disaster            221
Questionable             20
Boat                      5
Unknown                   1
Unconfirmed               1
Unverified                1
Under investigation       1
Name: count, dtype: int64


In [24]:
# Calculate the mode of the 'Type' column
mode_type = df_mod['Type'].mode()[0]

# Replace nan values with the mode
df_mod['Type'].fillna(mode_type, inplace=True)

# Verify that there are no nan values in the 'Type' column
nan_count = df_mod['Type'].isna().sum()
print(f"Number of nan values in 'Type' after imputation: {nan_count}")

Number of nan values in 'Type' after imputation: 0



### Analysis of the Country Column

In this section, we will analyze the `Country` column of the dataset. We will examine the different countries where incidents have been recorded, identify any anomalies, and clean the data as necessary. This analysis will help us understand the geographical distribution of the incidents and prepare the data for further analysis and visualization.

In [25]:
# Extract unique values from the 'Country' column
unique_countries = df_mod['Country'].unique()
print(f"Unique countries: {unique_countries}")


Unique countries: ['Morocco' 'Jamaica' 'Belize' 'Australia' 'USA' 'Maldive Islands'
 'Turks and Caicos' 'French Polynesia' 'Tobago' 'Bahamas' 'AUSTRALIA'
 'INDIA' 'TRINIDAD' 'BAHAMAS' 'SOUTH AFRICA' 'MEXICO' 'NEW ZEALAND'
 'EGYPT' 'Mexico' 'BELIZE' 'Coral Sea' 'SPAIN' 'PORTUGAL' 'SAMOA'
 'COLOMBIA' 'ECUADOR' 'CUBA' 'BRAZIL' 'SEYCHELLES' 'NEW CALEDONIA'
 'ARGENTINA' 'FIJI' 'MeXICO' 'Maldives' 'South Africa' 'ENGLAND' 'JAPAN'
 'INDONESIA' 'JAMAICA' 'MALDIVES' 'FRENCH POLYNESIA' 'THAILAND' 'COLUMBIA'
 'COSTA RICA' 'New Zealand' 'British Overseas Territory' 'CANADA' 'JORDAN'
 'ST KITTS / NEVIS' 'ST MARTIN' 'PAPUA NEW GUINEA' 'REUNION ISLAND'
 'ISRAEL' 'CHINA' 'IRELAND' 'ITALY' 'MALAYSIA' 'LIBYA' nan 'MAURITIUS'
 'SOLOMON ISLANDS' 'ST HELENA, British overseas territory' 'COMOROS'
 'REUNION' 'UNITED KINGDOM' 'UNITED ARAB EMIRATES' 'PHILIPPINES'
 'CAPE VERDE' 'Fiji' 'DOMINICAN REPUBLIC' 'CAYMAN ISLANDS' 'ARUBA'
 'MOZAMBIQUE' 'PUERTO RICO' 'ATLANTIC OCEAN' 'GREECE' 'ST. MARTIN'
 'FRANCE' 'TRIN

In [26]:
def clean_country(country):
    if pd.isna(country):
        return country
    
    # Strip leading and trailing whitespace
    country = country.strip()
    
    # Convert to title case
    country = country.title()
    
    # Remove special characters
    country = re.sub(r'[^\w\s]', '', country)
    
    return country

# Apply the cleaning function to the 'Country' column
df_mod['Country'] = df_mod['Country'].apply(clean_country)

# Verify the transformations
print(df_mod['Country'].head(10))

0      Morocco
1      Jamaica
2       Belize
3    Australia
4    Australia
5          Usa
6          Usa
7          Usa
8          Usa
9          Usa
Name: Country, dtype: object


In [27]:
# Extract unique values from the 'Country' column again
unique_countries_cleaned = df_mod['Country'].unique()
print(f"Cleaned unique countries: {unique_countries_cleaned}")


Cleaned unique countries: ['Morocco' 'Jamaica' 'Belize' 'Australia' 'Usa' 'Maldive Islands'
 'Turks And Caicos' 'French Polynesia' 'Tobago' 'Bahamas' 'India'
 'Trinidad' 'South Africa' 'Mexico' 'New Zealand' 'Egypt' 'Coral Sea'
 'Spain' 'Portugal' 'Samoa' 'Colombia' 'Ecuador' 'Cuba' 'Brazil'
 'Seychelles' 'New Caledonia' 'Argentina' 'Fiji' 'Maldives' 'England'
 'Japan' 'Indonesia' 'Thailand' 'Columbia' 'Costa Rica'
 'British Overseas Territory' 'Canada' 'Jordan' 'St Kitts  Nevis'
 'St Martin' 'Papua New Guinea' 'Reunion Island' 'Israel' 'China'
 'Ireland' 'Italy' 'Malaysia' 'Libya' nan 'Mauritius' 'Solomon Islands'
 'St Helena British Overseas Territory' 'Comoros' 'Reunion'
 'United Kingdom' 'United Arab Emirates' 'Philippines' 'Cape Verde'
 'Dominican Republic' 'Cayman Islands' 'Aruba' 'Mozambique' 'Puerto Rico'
 'Atlantic Ocean' 'Greece' 'France' 'Trinidad  Tobago' 'Kiribati'
 'Diego Garcia' 'Taiwan' 'Palestinian Territories' 'Guam' 'Nigeria'
 'Tonga' 'Scotland' 'Croatia' 'Saudi Arab

In [28]:
# Cheking for missing values in the 'Country' column
country_null_count = df_mod['Country'].isnull().sum()
print(f"Number of missing values in 'Country': {country_null_count}")

Number of missing values in 'Country': 43


In [29]:
# Calculate the % of missing values in the 'Country' column
country_null_percent = country_null_count / len(df_mod) * 100
print(f"Percentage of missing values in 'Country': {country_null_percent:.2f}%")


Percentage of missing values in 'Country': 0.66%


In [30]:
# Calculate the mode of the 'Country' column
mode_country = df_mod['Country'].mode()[0]

# Replace missing values with the mode
df_mod['Country'].fillna(mode_country, inplace=True)

# Verify that there are no missing values in the 'Country' column
country_null_count_after = df_mod['Country'].isnull().sum()
print(f"Number of missing values in 'Country' after imputation: {country_null_count_after}")


Number of missing values in 'Country' after imputation: 0


### Analysis of the Country Column

In this section, we will analyze the `State` column of the dataset. We will examine the different states recorded, identify any anomalies, and clean the data as necessary. This analysis will help us understand the geographical distribution of incidents within countries and prepare the data for more detailed analysis and visualization.


In [31]:
# Extract unique values from the 'State' column
unique_states = df_mod['State'].unique()
print(f"Unique states: {unique_states}")

Unique states: ['Southern Morocco' 'Montego Bay' 'Gulf of Honduras' 'NSW'
 'Western Australia' 'Florida' 'Texas' 'Floria' 'California'
 'Maahvah Laamu Atoll' 'Galveston' 'Providenciales Island'
 'Gambier Islands' 'South Carolina' 'Graet Courland Bay'
 'Clarence Town Long Isand' 'South Australia' 'Nunu Atoll' 'Queensland'
 'Hawaii' 'Maharashtra' nan 'New  South Wales' 'Paradise Island'
 'Eastern Cape Province' 'Sonora' 'New South Wales' 'South Island'
 'New Providence   Isoad' 'Jalisco' 'Grand  Bahama Island'
 'Westerm Australia' 'North Island' 'Southern Red Sea' 'Quintana Roo'
 'Belize District' 'South Sinai' 'North Carolina' 'Valencia' 'Norte'
 "Savai'i island" 'New York' 'San Andrés Island' 'Galapagos Islands'
 'Freeport' 'Lucayan Lucayan Archipelago' ' Utah' 'Mayabeque'
 'Easten Cape Province' 'Artemisa' 'Pernambuco' 'Praslin Island'
 'South Province' 'Patagonia' 'Taveuni Island' 'Gulf of California'
 'Vaavu Atoll' 'KNZ' 'New South ales' 'Samoa' 'Louisiana' 'Mississippi'
 'Western  

In [33]:
pf["State"].unique()

array(['Southern Morocco', 'Montego Bay', 'Gulf of Honduras', 'NSW',
       'Western Australia', 'Florida', 'Texas', 'Floria', 'California',
       'Maahvah Laamu Atoll', 'Galveston', 'Providenciales Island',
       'Gambier Islands', 'South Carolina', 'Graet Courland Bay',
       'Clarence Town Long Isand', 'South Australia', 'Nunu Atoll',
       'Queensland', 'Hawaii', 'Maharashtra', nan, 'New  South Wales',
       'Paradise Island', 'Eastern Cape Province', 'Sonora',
       'New South Wales', 'South Island', 'Guerrero',
       'New Providence   Isoad', 'Jalisco', 'Grand  Bahama Island',
       'Westerm Australia', 'North Island', 'Southern Red Sea',
       'Quintana Roo', 'Belize District', 'South Sinai', 'Bolinao',
       'North Carolina', 'Valencia', 'Norte', "Savai'i island",
       'New York', 'San Andrés Island', 'Galapagos Islands',
       'Red Sea Protectorate', 'Freeport', 'Tahiti', 'Poum',
       'Lucayan Lucayan Archipelago', ' Utah', 'New Jersey', 'Mayabeque',
       'Eas

In [34]:
pf["State"].nunique() #pendiente revision 

914

6 columna

In [35]:
pf["Location"]

0                            West of Dakhla
1                                  Falmouth
2                           Central America
3         North Shore Beach, Port Macquarie
4                      Trigg beach Sterling
                       ...                 
6965                            Roebuck Bay
6966                                    NaN
6967                         Ocracoke Inlet
6968                   Panama Bay 8ºN, 79ºW
6969    Below the English fort, Trincomalee
Name: Location, Length: 6970, dtype: object

In [36]:
pf["Location"].unique()

array(['West of Dakhla', 'Falmouth', 'Central America', ...,
       'Ocracoke Inlet', 'Panama Bay 8ºN, 79ºW',
       'Below the English fort, Trincomalee'], dtype=object)

In [37]:
pf["Location"].nunique()

4534

7 columna

In [38]:
pf["Activity"]

0       Swimming - jumped off yacht
1                      Spearfishing
2                Swimming - Diving 
3                           Surfing
4                           Surfing
                   ...             
6965                         Diving
6966                   Pearl diving
6967                       Swimming
6968                            NaN
6969                       Swimming
Name: Activity, Length: 6970, dtype: object

In [42]:
pf["Activity"].nunique()

1596

In [44]:
for i in pf["Activity"].unique(): 
    print(i)

Swimming - jumped off yacht
Spearfishing
Swimming - Diving 
Surfing
Diving into Water
Wading
Swimming
Playing football in the water
Swimming ocean training 
Military ocean training
snorkelling
Freediving
Jumpinf from dock to boat acidentally fell into water at marina
Fishing
nan
Free diving
Foil Boarding
Snorkeling
Paddle boarding
Scuba diving
On a round-the-world expedition
Kayak fishing
Towing a dead whale out to sea
Boogie boarding
Sitting
Watching the sardine run
Standing
Filming
Kite surfing
Jumped overboard
Shark Feeding
Diving for crayfish
Windsurfing
Kite-Surfing
Surf-sking
Diving with  surface-supplied air
Spearishing
Snorkeling at Shark Feeding site
Diving/Shark Feeding
Diving
Surf skiing 
Shipwreck
Swimming  or Snorkeling
Kayaking
Paddleboarding
Lobstering
Lifeguard Training Exercise
Lifeguard Exercises
Scalloping
Jumped into water
Playing in the water
Feeding sharks & stingrays
Fihing
Surface Supplied  Diving
Floating in inflatable pool ring
Fishing for sharks
Wing Foil Sur

8 columna

In [46]:
pf["Name"]

0              German Tourist
1                Jahmari Reid
2           Annabelle Carlson
3                Kai McKenzie
4                Ryan Lowther
                ...          
6965                     male
6966                    Ahmun
6967    Coast Guard personnel
6968          Jules Patterson
6969                     male
Name: Name, Length: 6970, dtype: object

In [48]:
pf["Name"].unique()

array(['German Tourist', 'Jahmari Reid', 'Annabelle Carlson', ...,
       'A "chiefess"', 'Ahmun', 'Jules Patterson'], dtype=object)

In [49]:
pf["Name"].nunique() #no son todos distintos y deberian de serlo

5708

9 columna

In [50]:
pf["Sex"]

0       F
1       M
2       F
3       M
4       M
       ..
6965    M
6966    M
6967    M
6968    M
6969    M
Name: Sex, Length: 6970, dtype: object

In [51]:
pf["Sex"].unique() #reagrupar en 2

array(['F', 'M', nan, ' M', 'M ', 'lli', 'M x 2', 'N', '.'], dtype=object)

In [52]:
pf["Sex"].nunique()

8

10 columna

In [54]:
pf["Age"]

0        30
1        16
2        15
3        23
4        41
       ... 
6965    NaN
6966    NaN
6967    NaN
6968    NaN
6969     15
Name: Age, Length: 6970, dtype: object

In [55]:
pf["Age"].unique() #pasar a numero o rango para poder hacer media de edad 

array([30, 16, 15, 23, 41, 14, 26, 'Middle age', 18, 21, nan, 45, 46, 19,
       27, 20, 65, 64, 24, '?', '13', '11', '46', '32', '29', '10', '64',
       '62', '39', '22', '15', '16', '21', '66', '52', '20', '44', '26',
       '47', '55', '30', '59', '8', '50', '45', '34', '38', '30s', '37',
       '9', '19', '20/30', '35', '65', '20s', '77', '60', 38, '49', '42',
       '!2', '24', '73', '25', '18', '50s', '58', '67', '17', '14', '6',
       '41', '53', '68', '43', '40', '51', '31', 39, 58, 51, 17, 10, 13,
       33, 40, 49, 60, 28, '40s', 68, 35, 62, 'teen', 8, 22, 32, 56,
       'Teen', 12, 42, 36, 37, 50, 'M', 9, 11, 57, 34, 25, '!6', 31, '!!',
       47, 55, 7, 71, 48, 59, 53, 54, 75, '45 and 15', 61, 73, 52, 29, 70,
       4, 63, 44, '28 & 22', '22, 57, 31', '60s', "20's", 43, 67, 74,
       '9 & 60', 'a minor', 6, 69, 3, 82, 66, 72, '23', '12', '36', '63',
       '71', '48', '70', '18 months', '57', '7', '28', '33', '61', '74',
       '27', '3', '56', '28 & 26', '5', '54', '86'

11

In [56]:
pf["Injury"]

0                                             Leg severed
1                          Head right arm and leg severed
2            Right lower leg amputated after serious bite
3                                   Serious injury to leg
4                          Minor injury to lower left leg
                              ...                        
6965                                                FATAL
6966                                                FATAL
6967                                                FATAL
6968                                                FATAL
6969    FATAL. "Shark bit him in half, carrying away t...
Name: Injury, Length: 6970, dtype: object

In [59]:
for i in pf["Injury"].unique():
    print(i)

Leg severed
Head right arm and leg severed
Right lower leg amputated after serious bite
Serious injury to leg
Minor injury to lower left leg
Lower left leg injury
Minor injury to left foot
Bite to left leg calf muscle removed
Minor cuts to calf
Superficial wounds to leg
Minor injury to right foot four tendons ruptured
Bite to left leg  
Significant injuries to pelvic and abdominal region
Significant bites to torso left arm and hand
Wounds to lower extremities Fatal
Bitten on left hand by 4-5ft shark
Major damage to right leg later amputated lower left leg left hand and forearm
Bite left forearm several ripped tendons severed nerves deep muscle wounds
Bitten on left arm
Bite wounds to left arm and leg wounds to abdomen and left hand
Bitten on left leg puncture wounds to kneecap
Minor bite to left leg
Minor injuries left foot
Minor injury to left leg
Minor injuries to back, abdomen and legs
No injury, shark bit surfboard
Lacerations to left foot
Leg bitten
Calf of lower left leg injured


In [58]:
pf["Injury"].nunique() #revisar datos validos

4100

12

In [60]:
pf["Unnamed: 11"]

0       Y
1       Y
2       N
3       N
4       N
       ..
6965    Y
6966    Y
6967    Y
6968    Y
6969    Y
Name: Unnamed: 11, Length: 6970, dtype: object

In [61]:
pf["Unnamed: 11"].unique() #posible a eliminar

array(['Y', 'N', nan, 'M', 'F', 'n', 'Nq', 'UNKNOWN', 2017, 'Y x 2', ' N',
       'N ', 'y'], dtype=object)

In [62]:
pf["Unnamed: 11"].nunique()

12

In [17]:
pf_mod.drop(columns=["Unnamed: 11"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Time            3443 non-null   object 
 12  Species         3838 non-null   object 
 13  Source          6951 non-null   object 
 14  original order  6799 non-null   float64
dtypes: float64(2), object(13)
memory usage: 816.9+ KB


13

In [63]:
pf["Time"]

0        1600hr
1           NaN
2       11.30hr
3        1100hr
4        1735hr
         ...   
6965        NaN
6966        NaN
6967        NaN
6968        NaN
6969        NaN
Name: Time, Length: 6970, dtype: object

In [64]:
pf["Time"].unique() #mismo formato horaa (plantear rango) y reagrupar los no conocidos

array(['1600hr', nan, '11.30hr', '1100hr', '1735hr', '11hr15', '16hr15',
       '?', '13h15', '9h', 1300, '14h', '15h30', 'Not stated', '13h30',
       '9h15', 'Not advised', '13h40', '12h30', '16h00', '11h30', '06h30',
       '20h00', '13h00', '11h12', '16h30', '15h00', '02h00', '09h15',
       'Early Morning', '16h32', '11h00', 'Morning', '10h30', '13h20',
       '14h00', '09h00', '10h20', '15h05', '17h00', '15h45', '07h45',
       '10h40', '07h50', '01h00', '10h00', 'Afternoon', '19h30',
       'Evening', '17h50', '09h30', '08h45', '"Midday"', '16h25', '13h55',
       '13h50', '17h20', '13h45', '10h10', '14h35', 'Night', '1500h ',
       '19h15', '11h20', '07h15', '07h00', '18h00', '08h00', '14h20',
       '17h30', '07h20', '14h50', '-16h30', '12h00', '17h17', '11h15',
       '19h00', '07h53', '16h10', '11h17', '17h45', '10jh45',
       'Early  morning', '13h12', '07h30', '11hoo', '11h43', '10h15',
       '14h09', '12h15', '19h12', '15h20', '16h40', '11h24', '12h50',
       '07h31',

14

In [65]:
pf["Species "]

0             Reportedly a Great White
1               Reportedly Tiger Shark
2                Reef shark 6ft (1.8m)
3                       3m White Shark
4       Undetermined small shall shark
                     ...              
6965                               NaN
6966                               NaN
6967                               NaN
6968                               NaN
6969                               NaN
Name: Species , Length: 6970, dtype: object

In [67]:
pf["Species "].nunique()

1693

In [68]:
pf["Species "].unique() #reestructurar

array(['Reportedly a Great White', 'Reportedly Tiger Shark',
       'Reef shark 6ft (1.8m)', ..., "12' tiger shark", 'Blue pointers',
       'Said to involve a grey nurse shark that leapt out of the water and  seized the boy but species identification is questionable'],
      dtype=object)

15

In [69]:
pf["Source"]

0              Andy Currie: Moroccan World News: Sky News
1        Todd Smith: Daily Mail UK: Sky News: People .com
2       Kevin McMurray Tracking Sharks.com New York Po...
3               Nine.com.au: ABC.Net.Au Johannes Marchand
4               Daily Mail: Sky News: The West Australian
                              ...                        
6965                     H. Taunton; N. Bartlett,  p. 234
6966                H. Taunton; N. Bartlett,  pp. 233-234
6967                 F. Schwartz, p.23; C. Creswell, GSAF
6968                                  The Sun, 10/20/1938
6969                                           S.W. Baker
Name: Source, Length: 6970, dtype: object

In [70]:
pf["Source"].unique() #posible a eliminar

array(['Andy Currie: Moroccan World News: Sky News',
       'Todd Smith: Daily Mail UK: Sky News: People .com',
       'Kevin McMurray Tracking Sharks.com New York Post: ABC News: The Independent:',
       ..., 'F. Schwartz, p.23; C. Creswell, GSAF', 'The Sun, 10/20/1938',
       'S.W. Baker'], dtype=object)

In [71]:
pf["Source"].nunique()

5326

- pdf                171
- href formula       176
- href               174
- Case Number        172
- Case Number.1      173
- original order     171
- Unnamed: 21       6969
- Unnamed: 22       6968

In [72]:
pf["pdf"] #eliminar

0                                  NaN
1                                  NaN
2                                  NaN
3                                  NaN
4                                  NaN
                     ...              
6965            ND-0005-RoebuckBay.pdf
6966                 ND-0004-Ahmun.pdf
6967    ND-0003-Ocracoke_1900-1905.pdf
6968        ND-0002-JulesPatterson.pdf
6969                ND-0001-Ceylon.pdf
Name: pdf, Length: 6970, dtype: object

In [10]:
pf_mod.drop(columns=["pdf"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 22 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  href formula    6794 non-null   object 
 16  href            6796 non-null   object 
 17  Case Number     6798 non-null   o

In [73]:
pf["href"] #eliminar

0                                                     NaN
1                                                     NaN
2                                                     NaN
3                                                     NaN
4                                                     NaN
                              ...                        
6965    http://sharkattackfile.net/spreadsheets/pdf_di...
6966    http://sharkattackfile.net/spreadsheets/pdf_di...
6967    http://sharkattackfile.net/spreadsheets/pdf_di...
6968    http://sharkattackfile.net/spreadsheets/pdf_di...
6969    http://sharkattackfile.net/spreadsheets/pdf_di...
Name: href, Length: 6970, dtype: object

In [11]:
pf_mod.drop(columns=["href"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  href formula    6794 non-null   object 
 16  Case Number     6798 non-null   object 
 17  Case Number.1   6797 non-null   o

In [75]:
pf["href formula"] #eliminar

0                                                     NaN
1                                                     NaN
2                                                     NaN
3                                                     NaN
4                                                     NaN
                              ...                        
6965    http://sharkattackfile.net/spreadsheets/pdf_di...
6966    http://sharkattackfile.net/spreadsheets/pdf_di...
6967    http://sharkattackfile.net/spreadsheets/pdf_di...
6968    http://sharkattackfile.net/spreadsheets/pdf_di...
6969    http://sharkattackfile.net/spreadsheets/pdf_di...
Name: href formula, Length: 6970, dtype: object

In [12]:
pf_mod.drop(columns=["href formula"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  Case Number     6798 non-null   object 
 16  Case Number.1   6797 non-null   object 
 17  original order  6799 non-null   f

In [76]:
pf["Case Number"] #eliminar

0           NaN
1           NaN
2           NaN
3           NaN
4           NaN
         ...   
6965    ND.0005
6966    ND.0004
6967    ND.0003
6968    ND.0002
6969    ND.0001
Name: Case Number, Length: 6970, dtype: object

In [13]:
pf_mod.drop(columns=["Case Number"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  Case Number.1   6797 non-null   object 
 16  original order  6799 non-null   float64
 17  Unnamed: 21     1 non-null      o

In [80]:
pf["original order"].unique() #que es???''

array([      nan, 6.802e+03, 6.801e+03, ..., 4.000e+00, 3.000e+00,
       2.000e+00])

In [81]:
pf["Case Number.1"] #eliminar

0           NaN
1           NaN
2           NaN
3           NaN
4           NaN
         ...   
6965    ND.0005
6966    ND.0004
6967    ND.0003
6968    ND.0002
6969    ND.0001
Name: Case Number.1, Length: 6970, dtype: object

In [14]:
pf_mod.drop(columns=["Case Number.1"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 18 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  original order  6799 non-null   float64
 16  Unnamed: 21     1 non-null      object 
 17  Unnamed: 22     2 non-null      o

In [82]:
pf["Unnamed: 21"] #eliminar

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
6965    NaN
6966    NaN
6967    NaN
6968    NaN
6969    NaN
Name: Unnamed: 21, Length: 6970, dtype: object

In [15]:
pf_mod.drop(columns=["Unnamed: 21"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  original order  6799 non-null   float64
 16  Unnamed: 22     2 non-null      object 
dtypes: float64(2), object(15)
memory 

In [83]:
pf["Unnamed: 22"] #eliminar

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
6965    NaN
6966    NaN
6967    NaN
6968    NaN
6969    NaN
Name: Unnamed: 22, Length: 6970, dtype: object

In [16]:
pf_mod.drop(columns=["Unnamed: 22"], inplace=True)
pf_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6970 entries, 0 to 6969
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            6970 non-null   object 
 1   Year            6968 non-null   float64
 2   Type            6952 non-null   object 
 3   Country         6920 non-null   object 
 4   State           6488 non-null   object 
 5   Location        6405 non-null   object 
 6   Activity        6384 non-null   object 
 7   Name            6750 non-null   object 
 8   Sex             6391 non-null   object 
 9   Age             3975 non-null   object 
 10  Injury          6935 non-null   object 
 11  Unnamed: 11     6408 non-null   object 
 12  Time            3443 non-null   object 
 13  Species         3838 non-null   object 
 14  Source          6951 non-null   object 
 15  original order  6799 non-null   float64
dtypes: float64(2), object(14)
memory usage: 871.4+ KB
